base/array.jl

```diff
@@ -1110,22 +1110,23 @@ function _growbeg_internal!(a::Vector, delta::Int, len::Int)
     if offset + len - 1 > memlen || offset < 1
         throw(ConcurrencyViolationError("Vector has invalid state. Don't modify internal fields incorrectly, or resize without correct locks"))
     end
-    # since we will allocate the array in the middle of the memory we need at least 2*delta extra space
+    # Since we will allocate the array in the middle of the memory we need at least 2*delta extra space
     # the +1 is because I didn't want to have an off by 1 error.
-    newmemlen = max(overallocation(len), len + 2 * delta + 1)
-    newoffset = div(newmemlen - newlen, 2) + 1
-    # If there is extra data after the end of the array we can use that space so long as there is enough
-    # space at the end that there won't be quadratic behavior with a mix of growth from both ends.
-    # Specifically, we want to ensure that we will only do this operation once before
-    # increasing the size of the array, and that we leave enough space at both the beginning and the end.
-    if newoffset + newlen < memlen
+    # Check if we can reuse existing memory without allocating
+    # This is especially important for empty arrays where overallocation(0) might be larger than memlen
+    newmemlen = len + 2 * delta + 1
+    if newmemlen <= memlen
```

Comment:
This isn't quite strict enough. We want to ensure that we only move the start of the array if there remains proportional space at both ends of the array (since otherwise you are signing up for potential quadratic behavior)

Reply:
Ah, you're right, I thought this seemed too easy. I changed it instead to use a different `overallocation` function logic. The current design does have a bit more "sawtooth" effect to it than we perhaps intended, so power of two sizes might grow much faster than the previous value (though in aggregate they grow over time equivalently):
```
r = 0:1024:1024^2; overallocation.(r) ./ r
```

Reply:
I wonder if we should have a `-1` somewhere in the formula. Making power of 2 be the breaking point might be pushing us away from optimal sizes (e.g. multiples of a page).